networkit/cpp/matching/BSuitorMatcher.cpp

```diff
@@ -0,0 +1,161 @@
+        std::istringstream istring(line);
+        int val;
+        if (!(istring >> val)) {
+            throw std::runtime_error("File " + path + " contains an invalid value in line "
+                                     + std::to_string(line_number) + ".");
+        }
+        if (istring >> val) {
+            throw std::runtime_error("File " + path + " contains multiple values in line "
+                                     + std::to_string(line_number) + ".");
+        }
+        if (val < 0) {
+            throw std::runtime_error("File " + path + " contains a negative value in line "
+                                     + std::to_string(line_number) + ".");
+        }
+        b.emplace_back(val);
+        line_number++;
+    }
+    if (b.size() != size) {
+        throw std::runtime_error("The number of values in file " + path
+                                 + " does not match the number of nodes in this graph.");
+    }
+    return b;
+}
+
+void BSuitorMatcher::findSuitors(node u) {
+    for (index i = 0; i < b.at(u); i++) {
+        auto [x, w] = findPreferred(u);
+        if (x != none) {
+            makeSuitor(u, w, x);
+        }
+    }
+}
+
+Node BSuitorMatcher::findPreferred(node y) {
+    Node best = Node{none, 0};
+
+    auto hasProposedTo = [&](node n) -> bool {
+        return std::any_of(T.at(y)->list.begin(), T.at(y)->list.end(),
+                           [n](const Node &p) { return p.id == n; });
+    };
+
+    for (auto v : G->weightNeighborRange(y)) {
+        const Node n = Node(v.first, v.second);
+        if (!hasProposedTo(n.id)) {
+            if (n.weight > best.weight || (n.weight == best.weight && n.id < best.id)) {
+                const auto n_suitor_weight = S.at(n.id)->min.weight;
+
+                if (n.weight > n_suitor_weight
+                    || (n.weight == n_suitor_weight && y < S.at(n.id)->min.id)) {
+                    best = n;
+                }
+            }
+        }
+    }
+    return best;
+}
+
+void BSuitorMatcher::makeSuitor(node u, edgeweight w, node x) {
+    auto y = S.at(x)->popMinIfFull();
+    S.at(x)->insert(Node(u, w));
+    T.at(u)->insert(Node(x, w));
+
+    if (y != none) {
+        T.at(y)->remove(Node(x, w));
+        auto [z, w] = findPreferred(y);
+        if (z != none) {
+            makeSuitor(y, w, z);
+        }
+    }
+}
+
+bool BSuitorMatcher::isSymmetrical() const {
+    bool sym = true;
+    auto areMatchedSymmetrical = [&](node u, node v) -> bool {
+        auto i_1 = std::find_if(S.at(u)->list.begin(), S.at(u)->list.end(),
+                                [v](const Node &p) { return p.id == v; })
+                   != S.at(u)->list.end();
+        auto i_2 = std::find_if(S.at(v)->list.begin(), S.at(v)->list.end(),
+                                [u](const Node &p) { return p.id == u; })
+                   != S.at(v)->list.end();
+        return (i_1 == i_2);
+    };
+
+    G->forNodes([&](node u) {
+        G->forNodes([&](node v) {
```

Comment:
Currently, every node pair is checked twice. You can omit this by only checking if `u > v`.

Reply:
Thanks.